ksql-engine/src/main/java/io/confluent/ksql/util/QueuedQueryMetadata.java

```diff
@@ -67,4 +74,12 @@ public boolean equals(Object o) {
   public int hashCode() {
     return Objects.hash(rowQueue, super.hashCode());
   }
+
+  private class StateListener implements KafkaStreams.StateListener {
+    @Override
+    public void onChange(final KafkaStreams.State newState, final KafkaStreams.State oldState) {
+      isRunning.set(newState == KafkaStreams.State.RUNNING
```

Comment:
if you really want to know when it is actually `RUNNING` you need to check that `oldState` was `REBALANCING` depends on what you are trying to achieve.
i.e., when the stream is started it transitions to `RUNNING`, but isn't really doing anything apart from subscribing, then goes to `REBALANCING`, then `RUNNING`

Reply:
Really, the main point is detect when its shutting down, so maybe just checking for newState being `NOT_RUNNING` would be best?

I'll switch to that.

Reply:
yeah - `NOT_RUNNING` in that case is better